File: analytics/src/test/java/io/appmetrica/analytics/impl/startup/StartupParamsAreResponseClidsConsistentTest.java
package io.appmetrica.analytics.impl.startup;

import android.content.Context;
import io.appmetrica.analytics.internal.IdentifiersResult;
import io.appmetrica.analytics.coreapi.internal.identifiers.IdentifierStatus;
import io.appmetrica.analytics.impl.db.preferences.PreferencesClientDbStorage;
import io.appmetrica.analytics.impl.utils.JsonHelper;
import io.appmetrica.analytics.impl.utils.StartupUtils;
import io.appmetrica.analytics.testutils.CommonTest;
import io.appmetrica.analytics.testutils.TestUtils;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.Map;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.ParameterizedRobolectricTestRunner;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.nullable;
import static org.mockito.Mockito.when;

@RunWith(ParameterizedRobolectricTestRunner.class)
public class StartupParamsAreResponseClidsConsistentTest extends CommonTest {

    private IdentifiersResult mInputResponseClids;
    private String mInputClientClids;
    private boolean mExpectedResult;

    public StartupParamsAreResponseClidsConsistentTest(Map<String, String> inputResponseClids,
                                                       Map<String, String> inputClientClids,
                                                       boolean expectedResult) {
        mInputResponseClids = new IdentifiersResult(JsonHelper.clidsToString(inputResponseClids), IdentifierStatus.OK, null);
        mInputClientClids = StartupUtils.encodeClids(inputClientClids);
        mExpectedResult = expectedResult;
    }

    @ParameterizedRobolectricTestRunner.Parameters(name = "[{index}]Return {2} for responseClids = \"{0}\" and " +
            "clientClids = \"{1}\"")
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][]{
                {null, null, true},
                {null, Collections.emptyMap(), true},
                {null, StartupParamsTestUtils.CLIDS_MAP_2, true},
                {Collections.emptyMap(), null, true},
                {Collections.emptyMap(), Collections.emptyMap(), true},
                {Collections.emptyMap(), StartupParamsTestUtils.CLIDS_MAP_2, false},
                {StartupParamsTestUtils.CLIDS_MAP_1, null, true},
                {StartupParamsTestUtils.CLIDS_MAP_1, Collections.emptyMap(), true},
                {StartupParamsTestUtils.CLIDS_MAP_1, StartupParamsTestUtils.CLIDS_MAP_2, true}
        });
    }

    private Context context;
    @Mock
    private PreferencesClientDbStorage mStorage;

    private StartupParams mStartupParams;

    @Before
    public void setUp() throws Exception {
        MockitoAnnotations.openMocks(this);
        context = TestUtils.createMockedContext();
        StartupParamsTestUtils.mockPreferencesClientDbStoragePutResponses(mStorage);

        when(mStorage.getClientClids(nullable(String.class))).thenReturn(mInputClientClids);
        when(mStorage.getResponseClidsResult()).thenReturn(mInputResponseClids);
        when(mStorage.getCustomSdkHosts()).thenReturn(new IdentifiersResult(null, IdentifierStatus.UNKNOWN, null));
        when(mStorage.getFeatures()).thenReturn(new FeaturesInternal(null, IdentifierStatus.UNKNOWN, null));

        mStartupParams = new StartupParams(context, mStorage);
    }

    @Test
    public void testAreResponseClidsConsistent() {
        assertThat(mStartupParams.areResponseClidsConsistent()).isEqualTo(mExpectedResult);
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/startup/StartupParamsAreResponseClidsConsistentTest.kt
package io.appmetrica.analytics.impl.startup

import io.appmetrica.analytics.coreapi.internal.identifiers.IdentifierStatus
import io.appmetrica.analytics.impl.db.preferences.PreferencesClientDbStorage
import io.appmetrica.analytics.impl.utils.JsonHelper
import io.appmetrica.analytics.impl.utils.StartupUtils
import io.appmetrica.analytics.internal.IdentifiersResult
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.ContextRule
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock

@RunWith(Parameterized::class)
internal class StartupParamsAreResponseClidsConsistentTest(
    inputResponseClids: MutableMap<String, String>?,
    inputClientClids: MutableMap<String, String>?,
    private val expectedResult: Boolean
) : CommonTest() {
    private val responseClids = IdentifiersResult(
        JsonHelper.clidsToString(inputResponseClids),
        IdentifierStatus.OK,
        null
    )
    private val clientClids: String = StartupUtils.encodeClids(inputClientClids)

    @get:Rule
    val contextRule = ContextRule()
    private val context by contextRule

    private val uuidResult = IdentifiersResult("test-uuid", IdentifierStatus.OK, null)

    private val storage: PreferencesClientDbStorage = mock {
        on { getClientClids(anyOrNull()) } doReturn clientClids
        on { responseClidsResult } doReturn responseClids
        on { customSdkHosts } doReturn IdentifiersResult(null, IdentifierStatus.UNKNOWN, null)
        on { getFeatures() } doReturn FeaturesInternal(null, IdentifierStatus.UNKNOWN, null)
        on { uuidResult } doReturn uuidResult
    }

    private lateinit var startupParams: StartupParams

    @Before
    fun setUp() {
        StartupParamsTestUtils.mockPreferencesClientDbStoragePutResponses(storage)
        startupParams = StartupParams(context, storage)
    }

    @Test
    fun areResponseClidsConsistent() {
        assertThat(startupParams.areResponseClidsConsistent()).isEqualTo(expectedResult)
    }

    companion object {

        @JvmStatic
        @Parameterized.Parameters(
            name = "[{index}]Return {2} for responseClids = \"{0}\" and " +
                "clientClids = \"{1}\""
        )
        fun data() = listOf(
            arrayOf(null, null, true),
            arrayOf(null, mutableMapOf<Any?, Any?>(), true),
            arrayOf(null, StartupParamsTestUtils.CLIDS_MAP_2, true),
            arrayOf(mutableMapOf<Any?, Any?>(), null, true),
            arrayOf(mutableMapOf<Any?, Any?>(), mutableMapOf<Any?, Any?>(), true),
            arrayOf(mutableMapOf<Any?, Any?>(), StartupParamsTestUtils.CLIDS_MAP_2, false),
            arrayOf(StartupParamsTestUtils.CLIDS_MAP_1, null, true),
            arrayOf(StartupParamsTestUtils.CLIDS_MAP_1, mutableMapOf<Any?, Any?>(), true),
            arrayOf(StartupParamsTestUtils.CLIDS_MAP_1, StartupParamsTestUtils.CLIDS_MAP_2, true)
        )
    }
}
